Add OpenTherm Gateway climate platform

mvn23 opened the pull request:
## Description:
Add support for the [OpenTherm Gateway](http://otgw.tclcode.com).

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6118

## Example entry for `configuration.yaml` (if applicable):
```yaml
climate:
  - platform: opentherm_gw
    device: /dev/ttyUSB0
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.i)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

mvn23 committed:
Initial OpenTherm Gateway support.

mvn23 committed:
Fix coveragerc and requirements_all*.txt
Overall cleanup and polishing

mvn23 committed:
Make hound/flake/travis happy

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+    @property
+    def min_temp(self):
+        """Return the minimum temperature."""
+        return convert_temperature(1, TEMP_CELSIUS, self.temperature_unit)
```
Since `temperature_unit` is celsius, we don't need to convert. Just `return 1`.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+    @property
+    def max_temp(self):
+        """Return the maximum temperature."""
+        return convert_temperature(30, TEMP_CELSIUS, self.temperature_unit)
```
See above.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+        """Set new target temperature."""
+        if 'temperature' in kwargs:
+            temp = float(kwargs['temperature'])
+            return self.gateway.set_target_temp(temp)
```
If `gateway.set_target_temp` is a coroutine we should await it here. Don't return anything.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+        return (getattr(self, '_away_state_a', False) or
+                getattr(self, '_away_state_b', False))
+
+    def async_set_temperature(self, **kwargs):
```
This should be a coroutine.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+    gateway.floor_temp = config.get(CONF_FLOOR_TEMP, False)
+    gateway.temp_precision = config.get(CONF_PRECISION)
+    async_add_entities([gateway])
+    await gateway.connect(hass, config.get(CONF_DEVICE))
```
Remove this. See comment below.

mvn23 replied:
Addressed in f6fb95fe6763bc9dac4b333e0c11a4c4ff85cfbb

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+"""OpenTherm Gateway Climate component for Home Assistant."""
```
This is a platform and not a component. Add a link to the docs. See other platforms for correct docstring format.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+"""OpenTherm Gateway Climate component for Home Assistant."""
+
```
Remove blank line.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+"""OpenTherm Gateway Climate component for Home Assistant."""
+
+import logging
+import voluptuous as vol
```
Add a blank line between standard library and 3rd party imports.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+                                              STATE_IDLE, STATE_HEAT,
+                                              STATE_COOL,
+                                              SUPPORT_TARGET_TEMPERATURE)
+from homeassistant.components.climate.modbus import CONF_PRECISION
```
Don't import things from other platforms.

mvn23 replied:
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

mvn23 commented:
Thank you for the very thorough review. Looks like there is still quite some work to do, but that was expected since it's my first real contribution to Home Assistant. I will address all concerns over the next few days.

mvn23 committed:
Basic improvements to comply with Home Assistant's style guidelines
Changed wording from "component" to more appropriate terms where necessary
Fixed small mistakes that snuck in during testing and/or due to my own ignorance ;)

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+    vol.Optional(CONF_FLOOR_TEMP, default=False): cv.boolean,
+})
+
+SUPPORT_FLAGS = (SUPPORT_TARGET_TEMPERATURE)
```
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4
I chose to keep the global variable in case more features will be added later on.

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+    await gateway.connect(hass, config.get(CONF_DEVICE))
+    _LOGGER.debug("Connected to %s on %s", gateway.friendly_name,
+                  config.get(CONF_DEVICE))
+    return True
```
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+
+    async def connect(self, hass, gw_path):
+        """Connect to the OpenTherm Gateway device at gw_path."""
+        self.hass = hass
```
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+        self.hass = hass
+        await self.gateway.connect(hass.loop, gw_path)
+        self.gateway.subscribe(self.receive_report)
+        return
```
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+    @property
+    def is_away_mode_on(self):
+        """Return true if away mode is on."""
+        return (getattr(self, '_away_state_a', False) or
```
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+
+    def async_set_temperature(self, **kwargs):
+        """Set new target temperature."""
+        if 'temperature' in kwargs:
```
addressed in c5a9266871b32a981ad19ab38a0bad650ab644e4

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,199 @@
+SUPPORT_FLAGS = (SUPPORT_TARGET_TEMPERATURE)
+_LOGGER = logging.getLogger(__name__)
+
+async def async_setup_platform(hass, config, async_add_entities,
```
Addressed in 728996927347a86923d94e33b8247d97fb4f1f36

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,199 @@
+                  config.get(CONF_DEVICE))
+    return True
+
+class OpenThermGateway(ClimateDevice):
```
Addressed in 728996927347a86923d94e33b8247d97fb4f1f36

mvn23 committed:
Fixed overwriting state property

mvn23 committed:
Fixed a bug with ROOM_SETPOINT_OVRD
Updated dependency pyotgw to latest version

mvn23 committed:
Remove unit_of_measurement from OpenThermGateway class

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+        """Return the precision of the system."""
+        if self.temp_precision is not None:
+            return self.temp_precision
+        if self.unit_of_measurement == TEMP_CELSIUS:
```
Addressed in 8fcedfada706ac193c0315cc03f30fb114f7082b

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+        return PRECISION_WHOLE
+
+    @property
+    def unit_of_measurement(self):
```
Addressed in 8fcedfada706ac193c0315cc03f30fb114f7082b

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+        return self.friendly_name
+
+    @property
+    def state(self):
```
Addressed in 704c725f9ef8020006ad24af52722f327bc5f2b0

mvn23 committed:
Cleanup after previous commits

mvn23 committed:
Moved initialisation and configuration from async_setup_platform to async_added_to_hass

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,185 @@
+        else:
+            self._current_operation = STATE_IDLE
+        self._current_temperature = status.get(self.pyotgw.DATA_ROOM_TEMP)
+        if status.get(self.pyotgw.DATA_ROOM_SETPOINT_OVRD) == None:
```
Addressed in 7195697644dc96a5283441563d2174f390aa4276

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,185 @@
+                                              STATE_IDLE, STATE_HEAT,
+                                              STATE_COOL,
+                                              SUPPORT_TARGET_TEMPERATURE)
+from homeassistant.const import (ATTR_TEMPERATURE, CONF_DEVICE, CONF_NAME,
```
Addressed in 7195697644dc96a5283441563d2174f390aa4276

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+        self._away_state_a = False
+        self._away_state_b = False
+
+    async def connect(self, hass, gw_path):
```
Addressed in f6fb95fe6763bc9dac4b333e0c11a4c4ff85cfbb

mvn23 commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,201 @@
+                               discovery_info=None):
+    """Set up the opentherm_gw component."""
+    gateway = OpenThermGateway()
+    gateway.friendly_name = config.get(CONF_NAME)
```
Addressed in f6fb95fe6763bc9dac4b333e0c11a4c4ff85cfbb

mvn23 committed:
Make travis happy

MartinHjelmare reviewed (commented):
Shouldn't polling be set to False for this entity?

mvn23 commented:
> Shouldn't polling be set to False for this entity?

I've been looking for a way to do this, but could not find it in the dev docs before. Seems I missed the "Entities Introduction" part and relied too much on examples from other platforms.
I will have a look and prepare another commit tonight or tomorrow.
Thanks for the feedback!

mvn23 committed:
Disable polling for this platform
Improve update flow

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,187 @@
+        import pyotgw
+        self.pyotgw = pyotgw
+        self.gateway = self.pyotgw.pyotgw()
+        self._device = config.get(CONF_DEVICE)
```
Use `dict[key]` for required config keys.

mvn23 replied:
Is this a Home Assistant specific requirement? I don't mind making the change, but a quick test confirmed that the performance impact is negligible.

MartinHjelmare replied:
It's not about performance. We should use appropriate tools for each task. When we know that a key is in a dictionary we access the value with `dict[key]`. If we would be mistaken, the code would blow up, and we would know about our mistake.

mvn23 replied:
Fair enough, addressed in 105f8eb668b11445d0de0862f32aebb783b9e3c1

MartinHjelmare commented on homeassistant/components/climate/opentherm_gw.py:
```diff
@@ -0,0 +1,187 @@
+        else:
+            self._current_operation = STATE_IDLE
+        self._current_temperature = status.get(self.pyotgw.DATA_ROOM_TEMP)
+        if status.get(self.pyotgw.DATA_ROOM_SETPOINT_OVRD) is None:
```
```py
temp = status.get(self.pyotgw.DATA_ROOM_SETPOINT_OVRD)
if temp is None:
    temp = status.get(self.pyotgw.DATA_ROOM_SETPOINT)
self._target_temperature = temp
```

mvn23 replied:
This optimization can be applied in other places as well. Will do in the next commit!

mvn23 replied:
Addressed in105f8eb668b11445d0de0862f32aebb783b9e3c1

MartinHjelmare reviewed (commented):
Nice! Two small comments.

mvn23 committed:
Small improvements/optimisations

MartinHjelmare reviewed (approved):
Great!
